documentation/DataModel.md

```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+
+### Preference Safes
+
+Preference Safes consist of a single primary document of type `prefsSafe`. These contain some optional metadata such
+as `name` and `email`, and the `preferences` section which contains the users preference sets. This is the central
+document for a safe, any documents relating to a safe should have a property `prefsSafeId` which contains the id of
+the preferences safe.
+
+An example document:
+
+```json
+    {
+        "_id": "prefsSafe-7",
+        "type": "prefsSafe",
+        "schemaVersion": "0.1",
+        "prefsSafeType": "user",
+        "name": null,
+        "email": null,
+        "preferences": {
+            "flat": {
+                "name": "bit of stuff",
+                "contexts": {
+                    "gpii-default": {
+                        "name": "Default preferences",
+                        "preferences": {
+                            "http://registry.gpii.net/common/onScreenKeyboard/enabled": true
+                        },
+                        "metadata": []
+                    }
+                },
+                "metadata": []
+            }
+        },
+        "timestampCreated": "2017-12-14T19:55:11.641Z",
+        "timestampUpdated": null
+    }
+```
+
+### Key-in Documents
+
+When users key-in to the GPII using a USB stick, NFC card, or other mechanism, the unique `gpiiKey` on the device will
+be matched to the CouchDB `_id` on a document with type `gpiiKey`. This document contains the important fields `prefsSafeId`
+and `prefsSetId` linking it to the safe, and to a specific preference set within that safe to key in with.
+
+An example document:
+
+```json
+    {
+        "_id": "np_tiny",
+        "type": "gpiiKey",
+        "schemaVersion": "0.1",
+        "prefsSafeId": "prefsSafe-7",
+        "prefsSetId": "gpii-default",
+        "revoked": false,
+        "revokedReason": null,
+        "timestampCreated": "2017-12-14T19:55:11.641Z",
+        "timestampUpdated": null
+    }
+```
+
+### Login Documents
+
+In order to have full permissions to edit all aspects of their preferences safe, users must login to their safe using a
+username and password. The current implementation of this is backed by the `gpii-express-user` library which creates
+records in the same format as native CouchDB accounts and manages password hashing, unlocking, etc.  In order to avoid
+making changes to this external library, we introduce a document type 'gpiiCloudSafeCredential' which tracks the native
+record that is created by `gpii-express-user`. Note that the `gpiiExpressuserId` entries are prefixed with `org.couch.db.user:`
+which is the convention for both internal CouchDB users and users created with gpii-express-user.
+
+An example document:
+
+```json
+    {
+        "type": "gpiiCloudSafeCredential",
+        "schemaVersion": "0.1",
+        "prefsSafeId": "prefsSafe-7",
+        "gpiiExpressUserId": "org.couch.db.user:prefs7user",
+        "_id": "8f3085a7-b65b-4648-9a78-8ac7de766997"
```

Comment:
Move the line for "_id" to the beginning of the JSON data block to match the convention above. The same comment applies to the JSON block below with "_id" at the bottom at line 112.

Reply:
This is updated.